File: 074/main.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int> >& matrix, int target) {
        if (!matrix.size()) return false;
        int row = matrix.size();
        int col = matrix[0].size();
        int start = 0;
        int end = row - 1;
        while (start < end - 1) {
            int mid = (start + end) / 2;
            if (matrix[mid][0] == target) return true;
            else if (matrix[mid][0] > target) end = mid;
            else start = mid;
        }
        for (int i = start; i <= end; i++) {
            for (int j = 0; j < col; j++) {
                if (matrix[i][j] == target) return true;
            }
        }
        return false;
    }
};

int main()
{
    Solution sol;
    vector<vector<int> > matrix;
    vector<int> m1;
    vector<int> m2;
    vector<int> m3;
    m1.push_back(1);
    m2.push_back(3);
    m3.push_back(5);
    matrix.push_back(m1);
    matrix.push_back(m2);
    matrix.push_back(m3);
    int res = sol.searchMatrix(matrix, 5);
    cout << res << endl;
    return 0;
}

File: 074/main.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  if (matrix == null || matrix.length === 0) return false
  const row = matrix.length
  const col = matrix[0].length
  let start = 0
  let end = row - 1
  while (start < end - 1) {
    let mid = Math.floor((start + end) / 2)
    if (matrix[mid][0] === target) return true
    else if (matrix[mid][0] > target) end = mid
    else start = mid
  }
  for (let i = start; i <= end; i++) {
    for (let j = 0; j <= col; j++) {
      if (matrix[i][j] === target) return true
    }
  }
  return false
}

console.log(searchMatrix([
  [1, 3, 5, 7],
  [10, 11, 16, 20],
  [23, 30, 34, 50]
]))
